### ISSUE TITLE
Queue weights aren't working as expected

### ISSUE THREAD

**Issue by sobrinho** [mperham/sidekiq#5741]
Ruby version: 2.7.6
Rails version: 6.0.3.2
Sidekiq / Pro / Enterprise version(s): Sidekiq 6.5.8 / Pro 5.5.5 / Ent 2.5.3

``` yaml
:concurrency: 4
:queues:
  - ["high", 20]
  - ["default", 6]
  - ["execupay", 4]
  - ["deluxe", 4]
  - ["gusto", 4]
  - ["paychex", 4]
  - ["heavy", 2]
  - ["segment",1]
```

We are observing that when `execupay` queue is under heavy load, `default` queue latency goes to like 6 hours even with higher weight:

<img width="679" alt="Screen Shot 2023-01-23 at 12 03 21" src="https://user-images.githubusercontent.com/26460/214073113-ee731aee-ffa3-42af-a265-6e43a7de5a22.png">

The funny part is that `deluxe` has the same weight with and had 1 hour of latency compared to 16 minutes on execupay.

What are we missing here?

**Comment by mperham** [comment]
Are you using superfetch or basicfetch?

**Comment by sobrinho** [comment]
super fetch

**Comment by sobrinho** [comment]
Are weights used only to pull jobs or does Sidekiq look at the running jobs to make sure what queue to pull next?

I'm guessing the first and since during that timeframe execupay jobs were lagging because unstable API (up to 2 minutes per job), that might be the cause.

Wondering what can be done to avoid that.

**Comment by mperham** [comment]
Yes, only the weights control which queue is fetched. The size of the queues is not relevant.

And so that screenshot doesn't really prove anything. Queues can be filled based on whatever has run, e.g. some code could enqueue a huge number of jobs in `default`. What you need to show is how the queues drain over time. The weights control the angle of the line as the queue drains. A large weight will drain quickly so the line will be very steep downward.

**Comment by sobrinho** [comment]
I see, I will add some instrumentation there.

The latency isn't indicating something is wrong? 6 hours of delay doesn't mean a job that was enqueued 6 hours ago is still there waiting to be processed?

**Comment by mperham** [comment]
It's a smell but not a smoking gun. You don't show the sizes of the other queues over time. How do you know the system wasn't working on 100,000 high jobs for the last 6 hours?

**Comment by sobrinho** [comment]
By using NewRelic.

The latency screenshot was taken around 10pm with 6 hours of delay in the default queue, so we can scope the timeframe to 4pm~10pm.

<img width="422" alt="Screen Shot 2023-01-23 at 14 39 05" src="https://user-images.githubusercontent.com/26460/214110294-868c006a-d5af-47da-98fa-663bf4102e05.png">

_Chart with 10min resolution._

Important to note that the jobs were struggling to drain due to the API instability, some of them was taking up to 10 minutes to error out, we are going to add more protection on those jobs against this to help.

But the question remains, aren't weights good enough to protect other queues from clogging like this?

**Comment by mperham** [comment]
> aren't weights good enough to protect other queues from clogging like this?

The only way to guarantee that one queue will drain before another queue 100% of the time is to use strictly ordered queues. Weighting is ultimately based on randomness. Bad job runtime due to that instability can cause all of your threads to be stuck on one type of job from one queue, slowing down the drain of that queue.

There absolutely could a bug here, it's just hard to tell.

**Comment by sobrinho** [comment]
I was wondering, the algorithm might look like this today:

```ruby
queues = ["default", "default", "low"] # assuming default:2,low:1

loop do
  queue = queues.sample
  job = pull(queue)
  process(job) if job
end
```

Why not track what's running in the current process to avoid the clogging? Maybe another specialized fetcher on Enterprise?

Not thinking of performance here yet, but:

```ruby
queues = ["default", "default", "low"]
desired_ratio = queues.tally # {"default"=>2, "low"=>1}
concurrency = 5
busy = Hash.new { |h, k| h[k] = 0 }

loop do
  queue = queues.sample

  # If we are going to exceed the desired ratio, do not pull from this queue
  if (concurrency.to_f / (busy[queue] + 1)) > desired_ratio[queue]
    next
  end

  job = pull(queue)

  if job
    # not quite like that since another thread is going to process the job
    # but imagine that the other thread would update our `busy` ivar
    busy[queue] += 1

    process(job)

    # and when it fails or succeeds, also update the busy
    busy[queue] -= 1
  end
end
```

_I'm not thinking of checking this across all the processes therefore using Redis, just the current process itself to avoid the roundtrip._

One drawback is that we won't use all the 5 threads if we have 1000 jobs on one queue and 0 in the other for a long period of time, but that's the intention behind this new fetcher.

In other words, I do not want ever to exceed the weight on busy, not from where to pull.

**Comment by sobrinho** [comment]
In terms of performance, we could be calculating the ratio across the different processes using Redis and updating a local cache for each process every few seconds, for instance.

So, there are ways of improving the ratio calculation but maybe it's not even needed, that's why I didn't think much about this yet.

**Comment by mperham** [comment]
I don't think this is the time or place to design a new fetcher. I've tried to keep the fetch algorithm simple and easy to understand. The complicated it gets, the more room there is for edge cases and bugs.

Can you reproduce your weighting issue with a small app?

**Comment by sobrinho** [comment]
I will try and submit here.

**Comment by sobrinho** [comment]
https://github.com/sobrinho/sidekiq_issue_5741

One way of reproducing, it just takes a while:

```
bin/rake sidekiq:issue_5741
bin/rails s -p 5000
bundle exec sidekiq -C config/sidekiq.yml
```

It starts just fine:

<img width="1840" alt="Screen Shot 2023-01-23 at 18 51 10" src="https://user-images.githubusercontent.com/26460/214158704-8a87ef4d-8b4d-4b8d-94f4-28a090d729fb.png">

Then after 2 minutes it's clogged:

<img width="1840" alt="Screen Shot 2023-01-23 at 18 53 50" src="https://user-images.githubusercontent.com/26460/214158779-54be53b7-60c7-4045-9ae9-bf333e9b0df1.png">

You can reproduce also instantly by doing in this order:

```
bundle exec sidekiq -C config/sidekiq.yml
bin/rails s -p 5000
bin/rake sidekiq:issue_5741
```

The problem here it is more obvious: while we are enqueueing the jobs, sidekiq pulls then and they just hang there.

<img width="1840" alt="Screen Shot 2023-01-23 at 18 49 12" src="https://user-images.githubusercontent.com/26460/214158877-249147c1-f7c8-40f2-ad03-6520ca013e75.png">

I used a sleep forever in the second job but since our jobs are taking about 10 minutes under heavy load doesn't matter since in about 2 minutes it's clogged.

**Comment by sobrinho** [comment]
Using the default fetcher it clogs after a few minutes.

Using the strict ordered queues isn't a option as well since it will clog with the first queue instead.

Using dedicated processes (as suggested in the wiki) isn't smart as well because we have 6 specialized queues where this might happen and it would require lots of memory incurring in high costs.

**Comment by sobrinho** [comment]
sidekiq-ent swarm isn't an option as well because it can't use different configuration for each process.

**Comment by mperham** [comment]
I'm not sure what you expect to happen. If your jobs all take a really long amount of time, yeah Sidekiq will get stuck because each worker thread fetches job and then works on it. If all the threads are stuck on long-running jobs, nothing will get fetched. Any network calls should have timeouts on them.

https://github.com/ankane/the-ultimate-guide-to-ruby-timeouts#the-ultimate-guide-to-ruby-timeouts

The behavior you show is exactly how I'd expect Sidekiq to behave. I can't generically add timeouts to jobs because `Timeout.timeout` is not safe to use in all cases.

**Comment by sobrinho** [comment]
I would like to be able to configure this behavior:

1. Concurrency: 10 threads
2. Queue A: occupy at most 8 threads
3. Queue B: occupy at most 2 threads

Having 100.000 jobs on Queue A would still occupy at most 8 threads and 2 threads would be idle waiting for work.

So, if I have a bad job on Queue A, I can still process Queue B jobs on the second thread.

The recommendation in the wiki is to have a separate process for this kind of queue but then I would need to spin up 6 dedicated processes for each queue consuming a large amount of resources (just to boot we need 2.5GB of RAM).

Also, I can't have the autoscaler starting from zero processes since the app takes about 3 minutes to boot and start processing.

>Using the weighted queues it clogs after a few minutes.
>
>Using the strict ordered queues isn't a option as well since it will clog with the first queue instead.
>
>Using dedicated processes (as suggested in the wiki) isn't smart as well because we have 6 specialized queues where this might happen and it would require lots of memory incurring in high costs.
>
>sidekiq-ent swarm isn't an option as well because it can't use different configuration for each process.

Ent Rate Limiter is also not an option since we autoscale therefore not having a exactly number of how many concurrent jobs we could have running.

--

There are some community initiatives such as https://github.com/deanpcmad/sidekiq-limit_fetch trying to solve this but they don't work with Sidekiq Pro/Ent.

**Comment by sobrinho** [comment]
But note, instead of saying a fixed number (because then I could be using limit fetch or rate limiter), we would using weight or a percentage of the process.

That way, I can scale up/down the processes/threads without worrying too much about this, it would readjust by itself.

That's why the pseudo-code here is converting the weight to percentage:

```ruby
queues = ["default", "default", "low"]
concurrency = 5

desired_ratio = queues.tally.transform_values { _1.to_f * 100 / queues.length }
# => {"default"=>66.66666666666667, "low"=>33.333333333333336}

busy = Hash.new { |h, k| h[k] = 0 }

loop do
  queue = queues.sample

  # If we are going to exceed the desired ratio, do not pull from this queue
  if ((busy[queue] + 1) * 100 / concurrency.to_f) > desired_ratio[queue]
    next
  end

  job = pull(queue)

  if job
    # not quite like that since another thread is going to process the job
    # but imagine that the other thread would update our `busy` ivar
    busy[queue] += 1

    process(job)

    # and when it fails or succeeds, also update the busy
    busy[queue] -= 1
  end
end
```

**Comment by sobrinho** [comment]
The math here is not perfect but good enough:

```ruby
queues = ["default", "default", "low"]
concurrency = 4
desired_ratio = queues.tally.transform_values { _1.to_f * 100 / queues.length }
# => {"default"=>66.66666666666667, "low"=>33.333333333333336}

max_threads = desired_ratio.map { [_1, _2 * concurrency / 100.0] }
#=> [["default", 2.666666666666667], ["low", 1.3333333333333335]]

max_threads = desired_ratio.map { [_1, (_2 * concurrency / 100.0).round] }
=> [["default", 3], ["low", 1]]
```

**Comment by mperham** [comment]
Sidekiq 7’s capsules allow you to set per-queue concurrency.

**Comment by sobrinho** [comment]
```ruby
Sidekiq.configure_server do |config|
  concurrency = config.concurrency
  ratios = config.queues.tally.transform_values { _1.to_f * 100 / config.queues.length }

  ratios.each do |queue, ratio|
    config.capsule(queue) do |capsule|
      capsule.queues = [queue]
      capsule.concurrency = (ratio * concurrency / 100.0).round
    end
  end
end
```

Unorthodox but seems to be working, I will give it a try.

**Comment by sobrinho** [comment]
A heads up that mutating the default capsule is changing `config.concurrency`, probably because it is the same as `config.capsules['default'].concurrency`.

That's why I'm copying the original value in the beginning of the block.

Not sure if the default capsule was supposed to be mutable but it worked here.

**Comment by mperham** [comment]
The only thing to understand is that those threads are dedicated to those queues. One nice thing about Sidekiq's traditional concurrency is that it will pick jobs from any queue so if you dump a huge number of jobs into `low` and all the other queues are empty, every Thread will pick up `low` jobs and drain the queue quickly.

With capsules, that doesn't happen. You limit the concurrency, which can be good or bad. That `low` queue won't drain quickly anymore.
